feat: sets up code in the new repo

nilekhc opened the pull request:
Fixes: https://github.com/kubernetes-sigs/secrets-store-sync-controller/issues/1

/kind feature

/label tide/merge-method-squash

AlexanderStocks commented on charts/secretsync/templates/_helpers.tpl:
```diff
@@ -0,0 +1,95 @@
+Determine the api version for the validating admission policies.
+*/}}
+{{- define "secret-sync-controller.admissionApiVersion" -}}
+{{- if semverCompare "~1.27.0" .Values.validatingAdmissionPolicies.kubernetesReleaseVersion -}}
```
~1.x.y will only match versions where the patch version increases. We should change this block to use >= and flip the cases so >= 1.29.0-0 comes first and then >=1.28.0-0 and so on. Then we can fall through the cases we want to catch and anything greater than 1.29 will not be classed as `unsupported-validating-admission-api-version`.

nilekhc replied:
We are using tilde (~) for version compare. It means any version that is compatible with for eg. 1.29.0. Hence in semverCompare any version that is greater than or equal to 1.29.0 and less than 2.0.0 is considered compatible.

AlexanderStocks reviewed (changes requested):
Bug

sozercan commented on Makefile:
```diff
@@ -0,0 +1,198 @@
+.PHONY: local-setup
+local-setup: docker-build ## setup and run sync controller locally
+	kind delete cluster --name sync-controller
+	kind create cluster --name sync-controller --image kindest/node:v1.29.2 --config=./hack/localsetup/kind-config.yaml
```
can you make the node version into a var?

sozercan commented on Makefile:
```diff
@@ -0,0 +1,198 @@
+# Update this value when you upgrade the version of your project.
+VERSION ?= v0.0.1
```
should this be 0.1.0?

nilekhc replied:
I'm thinking v0.0.1 for alpha, v0.1.0 for beta and v1.0.0 for GA. wdyt?

sozercan commented on Makefile:
```diff
@@ -0,0 +1,198 @@
+
+## Tool Versions
+KUSTOMIZE_VERSION ?= v4.5.7
+CONTROLLER_TOOLS_VERSION ?= v0.11.1
```
nit: can we move the versions to the top of the file so it's easy to track?

aramase reviewed (changes requested):
This is comments from 1st round of review for the non code files.

aramase commented on config/default/kustomization.yaml:
```diff
@@ -0,0 +1,76 @@
+# Adds namespace to all resources.
+namespace: secret-sync-controller-system
```
```suggestion
namespace: secrets-store-sync-controller-system
```

aramase commented on config/default/kustomization.yaml:
```diff
@@ -0,0 +1,76 @@
+# "wordpress" becomes "alices-wordpress".
+# Note that it should also match with the prefix (text before '-') of the namespace
+# field above.
+namePrefix: secret-sync-controller-
```
```suggestion
namePrefix: secrets-store-sync-controller-
```

aramase commented on config/manager/kustomization.yaml:
```diff
@@ -0,0 +1,8 @@
+kind: Kustomization
+images:
+- name: controller
+  newName: secret-sync-controller
```
```suggestion
  newName: secrets-store-sync-controller
```

aramase replied:
This comment applies to other places as well - let's use `secrets-store-sync-controller` where applicable

aramase commented on config/manager/manager.yaml:
```diff
@@ -0,0 +1,140 @@
+        control-plane: controller-manager
+    spec:
+      affinity:
+        nodeAffinity:
+          requiredDuringSchedulingIgnoredDuringExecution:
+            nodeSelectorTerms:
+            - matchExpressions:
+              - key: type
+                operator: NotIn
+                values:
+                - virtual-kubelet
```
This was added to the driver and has no relevance for the controller

aramase commented on config/manager/manager.yaml:
```diff
@@ -0,0 +1,140 @@
+      - operator: Exists
+      nodeSelector:
+        kubernetes.io/os: linux
+      #securityContext:
+        # runAsNonRoot: true
```
Remove these if not required

aramase commented on config/manifests/kustomization.yaml:
```diff
@@ -0,0 +1,26 @@
+resources:
+- bases/secret-sync-controller.clusterserviceversion.yaml
+- ../default
+- ../samples
```
This directory doesn't exist

aramase commented on config/manifests/kustomization.yaml:
```diff
@@ -0,0 +1,26 @@
+# These resources constitute the fully configured set of manifests
+# used to generate the 'manifests/' directory in a bundle.
+resources:
+- bases/secret-sync-controller.clusterserviceversion.yaml
```
This file doesn't exist

aramase commented on config/rbac/role.yaml:
```diff
@@ -0,0 +1,52 @@
+  - secrets
+  verbs:
+  - create
+  - delete
```
Why is `delete` permission required?

aramase commented on controllers/conditions.go:
```diff
@@ -0,0 +1,177 @@
+/*
+Copyright 2023.
```
This needs to be fixed!

aramase commented on controllers/conditions.go:
```diff
@@ -0,0 +1,177 @@
+*/
+package controllers
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	secretsyncv1alpha1 "sigs.k8s.io/secret-sync-controller/api/v1alpha1"
+)
```
Group imports

```go
import (
  stdlib

  external

  internal
)
```

comment applies to all the files

aramase commented on controllers/secretsync_controller.go:
```diff
@@ -0,0 +1,469 @@
+	Version = "v1"
+
+	// SecretSyncControllerFieldManager is the field manager used by the secret sync controller
+	SecretSyncControllerFieldManager = Version + "-secret-sync-controller"
```
```suggestion
	SecretSyncControllerFieldManager = Version + "-secrets-store-sync-controller"
```

aramase commented on controllers/secretsync_controller.go:
```diff
@@ -0,0 +1,469 @@
+
+//+kubebuilder:rbac:groups=secret-sync.x-k8s.io,resources=secretsyncs,verbs=get;list;watch
+//+kubebuilder:rbac:groups=secret-sync.x-k8s.io,resources=secretsyncs/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups="",resources=secrets,verbs=create;patch;delete
```
I don't think the `delete` verb is required, could you validate by removing it?

nilekhc replied:
I'll address this in the follow-up PR. Opened https://github.com/kubernetes-sigs/secrets-store-sync-controller/issues/16 for tracking.

aramase commented on controllers/secretsync_controller.go:
```diff
@@ -0,0 +1,469 @@
+	if err != nil {
+		logger.Error(err, "failed to get service account token", "name", ss.Spec.ServiceAccountName)
+
+		if checkIfErrorMessageCanBeDisplayed(err.Error()) {
+			r.updateStatusConditions(ctx, ss, ConditionTypeUnknown, conditionType, ConditionReasonValidatingAdmissionPolicyCheckFailed, true)
+		} else {
+			r.updateStatusConditions(ctx, ss, ConditionTypeUnknown, conditionType, ConditionReasonSecretPatchFailedUnknownError, true)
+		}
```
```suggestion
                conditionReason := ConditionReasonSecretPatchFailedUnknownError
		if checkIfErrorMessageCanBeDisplayed(err.Error()) {
			conditionReason = ConditionReasonValidatingAdmissionPolicyCheckFailed
		}
		
		r.updateStatusConditions(ctx, ss, ConditionTypeUnknown, conditionType, conditionReason, true)
```

aramase commented on controllers/secretsync_controller.go:
Could you add a unit test for this file? To start off in this PR, it would be good to have aa test for success and one that tests failures to assert this change is working. 

In a follow-up PR we can add table tests to validate all the possible scenarios.

nilekhc replied:
added tests as discussed

aramase commented on controllers/suite_test.go:
Remove this file if not used.

aramase commented on go.mod:
```diff
@@ -0,0 +1,86 @@
+module sigs.k8s.io/secret-sync-controller
```
```suggestion
module sigs.k8s.io/secrets-store-sync-controller
```

aramase commented on main.go:
```diff
@@ -0,0 +1,146 @@
+	"strings"
+	"time"
+
+	// Import all Kubernetes client auth plugins (e.g. Azure, GCP, OIDC, etc.)
+	// to ensure that exec-entrypoint and run can make use of them.
```
This comment is no longer relevant because the import has been dropped

aramase commented on docker/Dockerfile:
Moving the `Dockerfile` to the root dir is ok and we can remove the `docker/` dir

nilekhc replied:
I created the Dir so that we can place all the related files together. like `.dockerignore` and `Dockerfile`

aramase commented on docs/RELEASE.md:
Let's remove this file and open an issue to track adding this later

nilekhc replied:
Added https://github.com/kubernetes-sigs/secrets-store-sync-controller/issues/14 for tracking.

aramase commented on pkg/fakeprovider/fake_provider_client.go:
Where is this being used?

nilekhc replied:
removed

aramase reviewed (changes requested):
Added comments from another pass.

At the minimum, we'll need unit tests to ensure the reconcile loop is working and failing as expected as part of adding this PR. We can add more to the table tests in a follow-up.
- We'll need to add 2 jobs to start off (lint and unit-test), so we can perform those validations in this PR. cc @AlexanderStocks if you're interested in adding those jobs in test-infra.

nilekhc commented on .github/workflows/markdown-link-check.yaml:
```diff
@@ -0,0 +1,35 @@
+        use-quiet-mode: 'yes'
+        # this will show detailed HTTP status for checked links
+        use-verbose-mode: 'yes'
+        config-file: .markdownlinkcheck.json
```
updates with different action recommended by author

nilekhc commented on .github/workflows/publish-helm-chart.yaml:
```diff
@@ -0,0 +1,36 @@
+on:
+  push:
+    branches:
+      - main
```
yup, but if there are any changes pushed to a path `charts/**`. We have "path" configured for this.

```yaml
on:
  push:
    branches:
      - main
    paths:
      - ".github/workflows/publish-helm-chart.yaml"
      - "charts/**"
```

nilekhc commented on SECURITY.md:
It did not have any relevant content. Also, I did not find similar one in Secret Store CSI repo. So, to keep it consistent I deleted it for now. If we think we do have relevant content, then we could add it later, couldn't we?

nilekhc commented on .github/dependabot.yml:
```diff
@@ -0,0 +1,55 @@
+    commit-message:
+      prefix: "chore"
+    labels:
+      - "ok-to-test"
```
I added /hack/tools directory

nilekhc commented on PROJECT:
```diff
@@ -0,0 +1,19 @@
+layout:
+- go.kubebuilder.io/v3
+projectName: secret-sync-controller
+repo: sigs.k8s.io/secrets-store-csi-driver/secret-sync-controller
```
I'll addressed all the Kustomize related comments in the follow-up PR. Opened https://github.com/kubernetes-sigs/secrets-store-sync-controller/issues/15 for tracking

aramase commented on controllers/conditions.go:
```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2024.
```
```suggestion
Copyright 2024 The Kubernetes Authors.
```

aramase commented on go.mod:
```diff
@@ -0,0 +1,83 @@
+	k8s.io/klog/v2 v2.100.1
+	k8s.io/utils v0.0.0-20240310230437-4693a0247e57
+	sigs.k8s.io/controller-runtime v0.16.3
+	sigs.k8s.io/secrets-store-csi-driver v1.4.1
```
The fake provider in driver `v1.4.4` doesn't check for target path, so you can bump to that and remove the fake_server from here. This can be done in a follow-up.

aramase replied:
https://github.com/kubernetes-sigs/secrets-store-sync-controller/issues/18

aramase commented on go.mod:
```diff
@@ -0,0 +1,83 @@
+	go.opentelemetry.io/otel/exporters/prometheus v0.38.1
+	go.opentelemetry.io/otel/metric v0.38.1
+	go.opentelemetry.io/otel/sdk/metric v0.38.1
+	golang.org/x/crypto v0.17.0
```
Bump this to latest version as old versions have CVE

aramase replied:
https://github.com/kubernetes-sigs/secrets-store-sync-controller/issues/19 will help surface these before the release.

aramase commented on hack/tools/tools.go:
```diff
@@ -0,0 +1,24 @@
+// +build tools
+
+/*
+Copyright 2018 The Kubernetes Authors.
```
```suggestion
Copyright 2024 The Kubernetes Authors.
```

aramase commented on manifest_staging/charts/index.yaml:
Delete this file as the charts will be in `gh-pages`

aramase commented on manifest_staging/charts/secretsync/README.md:
```diff
@@ -0,0 +1,38 @@
+# Deploying the SecretSync Controller
```
In a follow-up let's change all the references to "Secrets Store Sync Controller"

aramase replied:
https://github.com/kubernetes-sigs/secrets-store-sync-controller/issues/20

aramase commented on pkg/k8s/token.go:
```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
```suggestion
Copyright 2024 The Kubernetes Authors.
```

aramase commented on pkg/k8s/token_test.go:
```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
```suggestion
Copyright 2024 The Kubernetes Authors.
```

aramase commented on pkg/metrics/exporter.go:
```diff
@@ -0,0 +1,27 @@
+package metrics
```
Add the license header

aramase commented on pkg/metrics/prometheus_exporter.go:
```diff
@@ -0,0 +1,52 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
```suggestion
Copyright 2024 The Kubernetes Authors.
```

aramase commented on pkg/provider/provider_client.go:
```diff
@@ -0,0 +1,309 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
```suggestion
Copyright 2024The Kubernetes Authors.
```

aramase commented on pkg/util/secretutil/secret.go:
```diff
@@ -0,0 +1,234 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
```suggestion
Copyright 2024 The Kubernetes Authors.
```

aramase reviewed (commented):
Added some comments. I've opened issues for follow-up's for most of them. Fix the other ones in this PR and we can merge it.

Before the `v0.0.1` release at the minimum we'll need:
1. The follow-up's you're working on for the RBAC permissions
2. e2e tests to validate the controller + VAPs we're adding
3. Failure unit test for controller

nilekhc commented:
> Added some comments. I've opened issues for follow-up's for most of them. Fix the other ones in this PR and we can merge it.
> 
> Before the `v0.0.1` release at the minimum we'll need:
> 
> 1. The follow-up's you're working on for the RBAC permissions
> 2. e2e tests to validate the controller + VAPs we're adding
> 3. Failure unit test for controller

Thanks for the review. I have addressed comments required in this PR. PTAL.
